File: 0992-subarrays-with-k-different-integers/0992-subarrays-with-k-different-integers.cpp
class Solution {
public:
    int subarraysWithKDistinct(vector<int>& nums, int k) {
        unordered_map<int,int> mp;
        int i=0,t=0,j=0;
        int ans=0;
        while(j<nums.size()){
            mp[nums[j]]++;

            while(mp.size()>k){
                mp[nums[t]]--;
                if(mp[nums[t]]==0)
                    mp.erase(nums[t]);
                t++;
                i=t;
            }

            while(mp[nums[t]]>1){
                mp[nums[t]]--;
                t++;
            }

            if(mp.size()==k)
                ans+=(t-i+1);
            j++;
        }
        return ans;
    }
};



/*
Map
1   ->  1
2   ->  2


i=far left
k=near left
j=right

1   2   1   2   3
i
        k
            j


result=1+2+
*/

File: 0992-subarrays-with-k-different-integers/0992-subarrays-with-k-different-integers.java
class Solution {
    public int subarraysWithKDistinct(int[] nums, int k) {
        int j=0;
        int n = nums.length;
        Map<Integer,Integer> mp = new HashMap<>();
        int[] expandedIndex = new int[n];
        for(int i=0;i<n;i++) {
            while(j<n && mp.size()<=k) {
                mp.put(nums[j],mp.getOrDefault(nums[j],0)+1);
                if(mp.size()>k) {mp.remove(nums[j]);break;}
                j++;
            }
            expandedIndex[i]=j;
            mp.put(nums[i],mp.get(nums[i])-1);
            if(mp.get(nums[i])==0) mp.remove(nums[i]);
        }
        j=0;
        mp.clear();
        int ans = 0;
        for(int i=0;i<n;i++) {
            while(j<n && mp.size()<k) {
                mp.put(nums[j],mp.getOrDefault(nums[j],0)+1);
                j++;
            }
            if(mp.size()==k) ans += expandedIndex[i]-j+1;
            mp.put(nums[i],mp.get(nums[i])-1);
            if(mp.get(nums[i])==0) mp.remove(nums[i]);
        }
        return ans;
    }
}

/*

0   1   2   3   4   5   6   7
1   2   1   1   2   3   2   1
i

3 + 3 + 3 + 3 + 1 + 1 = 14 


*/